### ISSUE TITLE
Forecasting Times: 3, 6, 12 months

### ISSUE THREAD

**Issue by jonathan-irvin** [jelly-fin/jelly-fin-web#11]
_From @jonathan-irvin on August 29, 2018 3:36_

Using the ability to generate transactions, we should be able to generate up to 12 months without committing them to the database.

_Copied from original issue: jelly-fin/jelly-fin#7_

**Comment by coder-log** [comment]
We will have different income and expenses behaviors. 

I could thing about these main behaviors:

Expenses:
- Linear/flat expenses with little variance. 
- Expenses with a seasonal behavior: energy for heating and cooling our homes for example, presents for christmas...
- Some of them will be impacted by inflation and interest rates, others not.
- Periodical expenses vs once in the life expenses.

Incomes:
- Salaries from jobs
- Bonus and extras
-  Income from assets: rents, dividends, bond interests...
- etc

**The main difficult to forecast any of them statiscally is that we need historic data to processs and a filter to evaluate and categorize it.**  

For new users the historic data will be missing (unless the user have a past registry -excel?- that could be imported in jelly-fish). 

On the other hand, the user could have an idea about the behavior of the income/expense (that also could be correct or not) so it could be a good idea to make the user categorize the perceived behavior 
 while the application begin tracking it on the shadows statistically in order to validate it.

So some logic must be applied:

`var historicData;
 if (historicData == True) {
     function evaluateForecastingMethod();
else 
    function userPerceivedForecastingMethod();
}
`

**Statistic Forecasting**
There are a bunch of techniques to forecast an historic serie, being the "Linear Regression" the most simple of them. 

For flat/linear expenses, this method will be the correct to apply. 

For the expenses that match these behavior the app might forecast a sole number (some kind of average) or an interval (applying 1x or 2x standard deviations).

For seasonal behavior a different formula must be used. I will elaborate on it next time.

The categorization itself could be intensive in resources if there are many users and historic data, so this could be a problem in the future and must be well thought.

**Comment by jonathan-irvin** [comment]
Yeah, I agree that we'd need past data to really establish accuracy.  Right now I base most of my stuff off of averages and get by alright.  We use a credit card for variable expenses and then pay it off at the end of the month.
